firebase-perf/src/main/java/com/google/firebase/perf/application/FrameMetricsRecorder.java

```diff
@@ -0,0 +1,148 @@
+package com.google.firebase.perf.application;
+
+import android.app.Activity;
+import android.util.SparseIntArray;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.core.app.FrameMetricsAggregator;
+
+import com.google.android.gms.common.util.VisibleForTesting;
+import com.google.firebase.perf.logging.AndroidLogger;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator;
+import com.google.firebase.perf.metrics.FrameMetricsCalculator.PerfFrameMetrics;
+import com.google.firebase.perf.util.Constants;
+
+import java.util.WeakHashMap;
+
+/**
+ * Provides FrameMetrics data from an Activity's Window. Encapsulates FrameMetricsAggregator.
+ * <p>
+ * Note: each recorder holds a reference to an Activity, so it is very important to dereference
+ * each recorder before the associated Activity is destroyed
+ * </p>
+ * */
+public class FrameMetricsRecorder {
+  private static final AndroidLogger logger = AndroidLogger.getInstance();
+  private final Activity activity;
+  private final FrameMetricsAggregator fma;
+  private final WeakHashMap<Object, PerfFrameMetrics> subTraceMap;
+  private boolean isRecording;
+
+  /**
+   * Creates a recorder for a specific activity
+   * @param activity the activity that the recorder is collecting data from
+   */
+  public FrameMetricsRecorder(Activity activity) {
+    this(activity, new FrameMetricsAggregator(), new WeakHashMap<>(), false);
+  }
+
+  @VisibleForTesting
+  FrameMetricsRecorder(Activity activity, FrameMetricsAggregator frameMetricsAggregator, WeakHashMap<Object, PerfFrameMetrics> subTraceMap, boolean isRecording) {
+    this.activity = activity;
+    this.fma = frameMetricsAggregator;
+    this.subTraceMap = subTraceMap;
+    this.isRecording = isRecording;
+  }
+
+  /**
+   * Starts recording FrameMetrics for the activity window
+   */
+  public void start() {
+    if (isRecording) {
+      logger.error("FrameMetricsAggregator is already recording %s", activity.getClass().getSimpleName());
+      return;
+    }
+    fma.add(activity);
+    isRecording = true;
+  }
+
+  /**
+   * Stops recording FrameMetrics for the activity window
+   * @return FrameMetrics accumulated during the current recording
+   */
+  public PerfFrameMetrics stop() {
+    if (!isRecording) {
+      logger.error("Cannot stop because no recording was started");
+      return null;
+    }
+    isRecording = false;
+    PerfFrameMetrics data = this.snapshot();
+    try {
+      fma.remove(activity);
+    } catch (IllegalArgumentException ignored) {
+      logger.error("View not hardware accelerated. Unable to collect FrameMetrics.");
```

Comment:
Should the hardware acceleration be checked before starting to record rather than wait for the stoping the recording?

Reply:
We used to have hardware acceleration checks but it was not working, so we deprecated it.